CryptoPkg/Library/BaseCryptLibMbedTls/Pk/CryptX509.c

```diff
@@ -0,0 +1,1939 @@
+  @param[out] SingleX509Cert  The generated X509 object.
+
+  @retval     TRUE            The X509 object generation succeeded.
+  @retval     FALSE           The operation failed.
+
+**/
+BOOLEAN
+EFIAPI
+X509ConstructCertificate (
+  IN CONST UINT8  *Cert,
+  IN UINTN        CertSize,
+  OUT UINT8       **SingleX509Cert
+  )
+{
+  mbedtls_x509_crt  *MbedTlsCert;
+  INT32             Ret;
+
+  if ((Cert == NULL) || (SingleX509Cert == NULL) || (CertSize == 0)) {
+    return FALSE;
+  }
+
+  MbedTlsCert = AllocateZeroPool (sizeof (mbedtls_x509_crt));
+  if (MbedTlsCert == NULL) {
+    return FALSE;
+  }
+
+  mbedtls_x509_crt_init (MbedTlsCert);
+
+  *SingleX509Cert = (UINT8 *)(VOID *)MbedTlsCert;
+  Ret             = mbedtls_x509_crt_parse_der (MbedTlsCert, Cert, CertSize);
+  if (Ret == 0) {
+    return TRUE;
+  } else {
+    mbedtls_x509_crt_free (MbedTlsCert);
+    FreePool (MbedTlsCert);
+    return FALSE;
+  }
+}
+
+/**
+  Construct a X509 stack object from a list of DER-encoded certificate data.
+
+  If X509Stack is NULL, then return FALSE.
+  If this interface is not supported, then return FALSE.
+
+  @param[in, out]  X509Stack  On input, pointer to an existing or NULL X509 stack object.
+                              On output, pointer to the X509 stack object with new
+                              inserted X509 certificate.
+  @param[in]       Args       VA_LIST marker for the variable argument list.
+                              A list of DER-encoded single certificate data followed
+                              by certificate size. A NULL terminates the list. The
+                              pairs are the arguments to X509ConstructCertificate().
+
+  @retval     TRUE            The X509 stack construction succeeded.
+  @retval     FALSE           The construction operation failed.
+  @retval     FALSE           This interface is not supported.
+
+**/
+BOOLEAN
+EFIAPI
+X509ConstructCertificateStackV (
+  IN OUT UINT8  **X509Stack,
+  IN VA_LIST    Args
+  )
+{
+  UINT8             *Cert;
+  UINTN             CertSize;
+  INT32             Index;
+  INT32             Ret;
+  mbedtls_x509_crt  *Crt;
+
+  if (X509Stack == NULL) {
+    return FALSE;
+  }
+
+  Ret = 0;
+  Crt = (mbedtls_x509_crt *)*X509Stack;
+  if (Crt == NULL) {
+    Crt = AllocateZeroPool (sizeof (mbedtls_x509_crt));
+    if (Crt == NULL) {
+      return FALSE;
+    }
+
+    mbedtls_x509_crt_init (Crt);
+    *X509Stack = (UINT8 *)Crt;
+  }
+
+  for (Index = 0; ; Index++) {
+    //
+    // If Cert is NULL, then it is the end of the list.
+    //
+    Cert = VA_ARG (Args, UINT8 *);
+    if (Cert == NULL) {
+      break;
+    }
+
+    CertSize = VA_ARG (Args, UINTN);
+    if (CertSize == 0) {
+      break;
+    }
+
+    Ret = mbedtls_x509_crt_parse_der (Crt, Cert, CertSize);
+
+    if (Ret != 0) {
+      break;
+    }
+  }
+
+  if (Ret == 0) {
+    return TRUE;
+  } else {
+    if ((mbedtls_x509_crt *)*X509Stack == NULL) {

```

Comment:
meaningless check, *X509Stack always be non-NULL when need to free.

Reply:
Yes. I have changed the logic. Thanks.